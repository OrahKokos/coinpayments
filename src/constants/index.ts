export const API_VERSION = 1;
export const API_PROTOCOL = 'https:';
export const API_HOST = 'www.coinpayments.net';
export const API_PATH = '/api.php';
export const API_FORMAT = 'json';
export const API_VALID_RESPONSE = 'ok';

export const CMDS = Object.freeze({
  RATES: 'rates',
  GET_BASIC_INFO: 'get_basic_info',
  CREATE_TRANSACTION: 'create_transaction',
  GET_CALLBACK_ADDRESS: 'get_callback_address',
  GET_TX: 'get_tx_info',
  GET_TX_MULTI: 'get_tx_info_multi',
  GET_TX_LIST: 'get_tx_ids',
  BALANCES: 'balances',
  GET_DEPOSIT_ADDRESS: 'get_deposit_address',
  CREATE_TRANSFER: 'create_transfer',
  CREATE_WITHDRAWAL: 'create_withdrawal',
  CREATE_MASS_WITHDRAWAL: 'create_mass_withdrawal',
  CANCEL_WITHDRAWAL: 'cancel_withdrawal',
  CONVERT: 'convert',
  CONVERT_LIMITS: 'convert_limits',
  GET_WITHDRAWAL_HISTORY: 'get_withdrawal_history',
  GET_WITHDRAWAL_INFO: 'get_withdrawal_info',
  GET_CONVERSATION_INFO: 'get_conversion_info',
  GET_TAG_INFO: 'get_pbn_info',
  GET_TAG_LIST: 'get_pbn_list',
  BUY_TAG: 'buy_pbn_tags',
  CLAIM_TAG: 'claim_pbn_tag',
  UPDATE_TAG: 'update_pbn_tag',
  RENEW_TAG: 'renew_pbn_tag',
  DELETE_TAG: 'delete_pbn_tag',
  CLAIM_COUPON: 'claim_pbn_coupon',
});
