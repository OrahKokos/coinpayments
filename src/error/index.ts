export default class CoinpaymentsError extends Error {
  public name: string;
  public message: string;
  public extra: Record<string, unknown>;

  constructor(message: string, extra: Record<string, unknown> = {}) {
    super(message);
    Error.captureStackTrace(this, CoinpaymentsError);
    this.message = message;
    this.extra = extra;
    Object.setPrototypeOf(this, new.target.prototype);
  }
}
