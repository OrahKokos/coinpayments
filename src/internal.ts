import { createHmac } from 'crypto';
import { request as httpsRequest } from 'https';
import { stringify } from 'querystring';

import {
  API_HOST,
  API_PATH,
  API_PROTOCOL,
  API_VERSION,
  API_FORMAT,
  API_VALID_RESPONSE,
} from './constants';
import CoinpaymentsError from './error';
import { validatePayload } from './validation';

import {
  CoinpaymentsCredentials,
  CoinpaymentsRequest,
  CoinpaymentsInternalRequestOps,
  CoinpaymentsInternalResponse,
  CoinpaymentsReturnCallback,
} from './types/base';

export const getPrivateHeaders = (
  credentials: CoinpaymentsCredentials,
  options: CoinpaymentsRequest,
) => {
  const { secret } = credentials;

  const paramString = stringify(options);
  const signature = createHmac('sha512', secret)
    .update(paramString)
    .digest('hex');

  return {
    'Content-Type': 'application/x-www-form-urlencoded',
    HMAC: signature,
  };
};

export const makeRequest = <ExpectedResponse>(
  reqOps: CoinpaymentsInternalRequestOps,
  options: CoinpaymentsRequest,
): Promise<ExpectedResponse> => {
  return new Promise((resolve, reject) => {
    const req = httpsRequest(reqOps, res => {
      let chunks = '';

      res.setEncoding('utf8');

      res.on('data', chunk => {
        chunks += chunk;
      });

      res.on('end', () => {
        let data: CoinpaymentsInternalResponse<ExpectedResponse> = {
          error: API_VALID_RESPONSE,
        };
        try {
          data = JSON.parse(chunks);
        } catch (e) {
          return reject(
            new CoinpaymentsError('Invalid response', { data: chunks }),
          );
        }

        if (data.error !== API_VALID_RESPONSE) {
          return reject(new CoinpaymentsError(data.error, { data }));
        }
        return resolve(data.result);
      });
    });
    req.on('error', reject);
    req.write(stringify(options));
    return req.end();
  });
};

export const getRequestOptions = (
  credentials: CoinpaymentsCredentials,
  options: CoinpaymentsRequest,
): CoinpaymentsInternalRequestOps => {
  return {
    protocol: API_PROTOCOL,
    method: 'post',
    host: API_HOST,
    path: API_PATH,
    headers: getPrivateHeaders(credentials, options),
  };
};

export const applyDefaultOptionValues = (
  credentials: CoinpaymentsCredentials,
  options: CoinpaymentsRequest,
): CoinpaymentsRequest => {
  return {
    ...options,
    version: API_VERSION,
    format: API_FORMAT,
    key: credentials.key,
  };
};

export const request = async <ExpectedResponse>(
  credentials: CoinpaymentsCredentials,
  options: CoinpaymentsRequest,
  callback?: CoinpaymentsReturnCallback<ExpectedResponse>,
): Promise<ExpectedResponse> => {
  try {
    validatePayload(options);
    options = applyDefaultOptionValues(credentials, options);
    const reqOps = getRequestOptions(credentials, options);
    const response: ExpectedResponse = await makeRequest<ExpectedResponse>(
      reqOps,
      options,
    );
    if (callback) {
      return callback(null, response);
    }
    return response;
  } catch (e) {
    if (callback) {
      return callback(e);
    }
    return Promise.reject(e);
  }
};
