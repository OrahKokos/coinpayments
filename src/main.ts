import Coinpayments from './index'

export default Coinpayments
export { Coinpayments }

module.exports = Coinpayments
