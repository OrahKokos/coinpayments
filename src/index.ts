import { CMDS } from './constants';
import CoinpaymentsError from './error';
import {
  mapPayload,
  mapMassWithdrawalPayload,
  mapGetTxMultiPayload,
} from './mappers';
import { request } from './internal';

import {
  CoinpaymentsCredentials,
  CoinpaymentsReturnCallback,
} from './types/base';
import {
  CoinpaymentsRatesOpts,
  CoinpaymentsGetBasicInfoOpts,
  CoinpaymentsGetCallbackAddressOpts,
  CoinpaymentsCreateTransactionOpts,
  CoinpaymentsGetTxOpts,
  CoinpaymentsGetTxMultiOpts,
  CoinpaymentsGetTxListOpts,
  CoinpaymentsBalancesOpts,
  CoinpaymentsGetDepositAddressOpts,
  CoinpaymentsCreateMassWithdrawalOpts,
  CoinpaymentsConvertCoinsOpts,
  CoinpaymentsConvertLimitsOpts,
  CoinpaymentsGetWithdrawalHistoryOpts,
  CoinpaymentsGetWithdrawalInfoOpts,
  CoinpaymentsCancelWithdrawalOpts,
  CoinpaymentsGetConversionInfoOpts,
  CoinpaymentsGetProfileOpts,
  CoinpaymentsTagListOpts,
  CoinpaymentsClaimTagOpts,
  CoinpaymentsUpdateTagProfileOpts,
  CoinpaymentsRenewTagOpts,
  CoinpaymentsDeleteTagOpts,
  CoinpaymentsClaimCouponOpts,
  CoinpaymentsBuyTagOpts,
  CoinpaymentsCreateTransferOpts,
  CoinpaymentsCreateWithdrawalOpts,
} from './types/options';

import {
  CoinpaymentsRatesResponse,
  CoinpaymentsCreateTransactionResponse,
  CoinpaymentsBalancesResponse,
  CoinpaymentsCreateWithdrawalResponse,
  CoinpaymentsCreateMassWithdrawalResponse,
  CoinpaymentsGetTxResponse,
  CoinpaymentsGetWithdrawalInfoResponse,
  CoinpaymentsGetTxListResponse,
  CoinpaymentsGetTxMultiResponse,
  CoinpaymentsGetBasicInfoResponse,
  CoinpaymentsGetDepositAddressResponse,
  CoinpaymentsGetCallbackAddressResponse,
  CoinpaymentsCreateTransferResponse,
  CoinpaymentsConvertCoinsResponse,
  CoinpaymentsConvertLimitsResponse,
  CoinpaymentsGetWithdrawalHistoryResponse,
  CoinpaymentsConversionInfoResponse,
  CoinpaymentsGetProfileResponse,
  CoinpaymentsTagListResponse,
  CoinpaymentsUpdateTagProfileResponse,
  CoinpaymentsClaimTagResponse,
  CoinpaymentsRenameTagResponse,
  CoinpaymentsDeleteTagResponse,
  CoinpaymentsClaimCouponResponse,
  CoinpaymentsBuyTagResponse,
} from './types/response';
import { filterMassWithdrawalOpts } from './validation';

class Coinpayments {
  private credentials: CoinpaymentsCredentials;

  constructor({ key = '', secret = '' }: CoinpaymentsCredentials) {
    if (!key) {
      throw new CoinpaymentsError('Missing public key');
    }
    if (!secret) {
      throw new CoinpaymentsError('Missing private key');
    }

    this.credentials = { key, secret };

    this.getBasicInfo = this.getBasicInfo.bind(this);
    this.rates = this.rates.bind(this);
    this.balances = this.balances.bind(this);
    this.getDepositAddress = this.getDepositAddress.bind(this);
    this.createTransaction = this.createTransaction.bind(this);
    this.getCallbackAddress = this.getCallbackAddress.bind(this);
    this.getTx = this.getTx.bind(this);
    this.getTxList = this.getTxList.bind(this);
    this.getTxMulti = this.getTxMulti.bind(this);
    this.createTransfer = this.createTransfer.bind(this);
    this.convertCoins = this.convertCoins.bind(this);
    this.convertLimits = this.convertLimits.bind(this);
    this.getWithdrawalHistory = this.getWithdrawalHistory.bind(this);
    this.getWithdrawalInfo = this.getWithdrawalInfo.bind(this);
    this.getConversionInfo = this.getConversionInfo.bind(this);
    this.getProfile = this.getProfile.bind(this);
    this.tagList = this.tagList.bind(this);
    this.updateTagProfile = this.updateTagProfile.bind(this);
    this.claimTag = this.claimTag.bind(this);
    this.cancelWithdrawal = this.cancelWithdrawal.bind(this);
    this.createWithdrawal = this.createWithdrawal.bind(this);
    this.createMassWithdrawal = this.createMassWithdrawal.bind(this);
  }

  public rates(
    options?:
      | CoinpaymentsRatesOpts
      | CoinpaymentsReturnCallback<CoinpaymentsRatesResponse>,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsRatesResponse>,
  ): Promise<CoinpaymentsRatesResponse> {
    if (!options && !callback) {
      options = {};
    }
    if (typeof options === 'function') {
      callback = options;
      options = {};
    }
    const requestPayload = mapPayload<CoinpaymentsRatesOpts>(options, {
      cmd: CMDS.RATES,
    });

    return request<CoinpaymentsRatesResponse>(
      this.credentials,
      requestPayload,
      callback,
    );
  }

  public createTransaction(
    options: CoinpaymentsCreateTransactionOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsCreateTransactionResponse>,
  ) {
    return request<CoinpaymentsCreateTransactionResponse>(
      this.credentials,
      mapPayload<CoinpaymentsCreateTransactionOpts>(options, {
        cmd: CMDS.CREATE_TRANSACTION,
      }),
      callback,
    );
  }

  public balances(
    options?:
      | CoinpaymentsBalancesOpts
      | CoinpaymentsReturnCallback<CoinpaymentsBalancesResponse>,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsBalancesResponse>,
  ) {
    if (typeof options === 'function') {
      callback = options;
      options = {};
    }
    return request<CoinpaymentsBalancesResponse>(
      this.credentials,
      mapPayload<CoinpaymentsBalancesOpts>(options, {
        cmd: CMDS.BALANCES,
      }),
      callback,
    );
  }

  public createWithdrawal(
    options: CoinpaymentsCreateWithdrawalOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsCreateWithdrawalResponse>,
  ) {
    options = { auto_confirm: 1, ...options };
    return request<CoinpaymentsCreateWithdrawalResponse>(
      this.credentials,
      mapPayload<CoinpaymentsCreateWithdrawalOpts>(options, {
        cmd: CMDS.CREATE_WITHDRAWAL,
      }),
      callback,
    );
  }

  public createMassWithdrawal(
    withdrawalArray: CoinpaymentsCreateMassWithdrawalOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsCreateMassWithdrawalResponse>,
  ) {
    return request<CoinpaymentsCreateMassWithdrawalResponse>(
      this.credentials,
      mapMassWithdrawalPayload(filterMassWithdrawalOpts(withdrawalArray), {
        cmd: CMDS.CREATE_MASS_WITHDRAWAL,
      }),
      callback,
    );
  }

  public cancelWithdrawal(
    options: CoinpaymentsCancelWithdrawalOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsCancelWithdrawalOpts>,
  ) {
    return request<CoinpaymentsCancelWithdrawalOpts>(
      this.credentials,
      mapPayload<CoinpaymentsCancelWithdrawalOpts>(options, {
        cmd: CMDS.CANCEL_WITHDRAWAL,
      }),
      callback,
    );
  }

  public getTx(
    options: CoinpaymentsGetTxOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsGetTxResponse>,
  ) {
    return request<CoinpaymentsGetTxResponse>(
      this.credentials,
      mapPayload<CoinpaymentsGetTxOpts>(options, {
        cmd: CMDS.GET_TX,
      }),
      callback,
    );
  }

  public getWithdrawalInfo(
    options: CoinpaymentsGetWithdrawalInfoOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsGetWithdrawalInfoResponse>,
  ) {
    return request<CoinpaymentsGetWithdrawalInfoResponse>(
      this.credentials,
      mapPayload<CoinpaymentsGetWithdrawalInfoOpts>(options, {
        cmd: CMDS.GET_WITHDRAWAL_INFO,
      }),
      callback,
    );
  }

  public getTxMulti(
    txIdArray: CoinpaymentsGetTxMultiOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsGetTxMultiResponse>,
  ) {
    if (!(txIdArray instanceof Array) || !txIdArray.length) {
      const error = new CoinpaymentsError('Invalid argument', { txIdArray });
      if (callback) {
        return callback(error);
      }
      return Promise.reject(error);
    }
    return request<CoinpaymentsGetTxMultiResponse>(
      this.credentials,
      mapGetTxMultiPayload(txIdArray, {
        cmd: CMDS.GET_TX_MULTI,
      }),
      callback,
    );
  }

  public getTxList(
    options?:
      | CoinpaymentsGetTxListOpts
      | CoinpaymentsReturnCallback<CoinpaymentsGetTxListResponse>,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsGetTxListResponse>,
  ) {
    if (!options && !callback) {
      options = {};
    }
    if (typeof options === 'function') {
      callback = options;
      options = {};
    }
    return request<CoinpaymentsGetTxListResponse>(
      this.credentials,
      mapPayload<CoinpaymentsGetTxListOpts>(options, {
        cmd: CMDS.GET_TX_LIST,
      }),
      callback,
    );
  }

  public getBasicInfo(
    callback?: CoinpaymentsReturnCallback<CoinpaymentsGetBasicInfoResponse>,
  ) {
    return request<CoinpaymentsGetBasicInfoResponse>(
      this.credentials,
      mapPayload<CoinpaymentsGetBasicInfoOpts>(
        {},
        {
          cmd: CMDS.GET_BASIC_INFO,
        },
      ),
      callback,
    );
  }

  public getDepositAddress(
    options: CoinpaymentsGetDepositAddressOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsGetDepositAddressResponse>,
  ) {
    return request<CoinpaymentsGetDepositAddressResponse>(
      this.credentials,
      mapPayload<CoinpaymentsGetDepositAddressOpts>(options, {
        cmd: CMDS.GET_DEPOSIT_ADDRESS,
      }),
      callback,
    );
  }

  public getCallbackAddress(
    options: CoinpaymentsGetCallbackAddressOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsGetCallbackAddressResponse>,
  ) {
    return request<CoinpaymentsGetCallbackAddressResponse>(
      this.credentials,
      mapPayload<CoinpaymentsGetCallbackAddressOpts>(options, {
        cmd: CMDS.GET_CALLBACK_ADDRESS,
      }),
      callback,
    );
  }

  public createTransfer(
    options: CoinpaymentsCreateTransferOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsCreateTransferResponse>,
  ) {
    return request<CoinpaymentsCreateTransferResponse>(
      this.credentials,
      mapPayload<CoinpaymentsCreateTransferOpts>(options, {
        cmd: CMDS.CREATE_TRANSFER,
        auto_confirm: true,
      }),
      callback,
    );
  }

  public convertCoins(
    options: CoinpaymentsConvertCoinsOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsConvertCoinsResponse>,
  ) {
    return request<CoinpaymentsConvertCoinsResponse>(
      this.credentials,
      mapPayload<CoinpaymentsConvertCoinsOpts>(options, {
        cmd: CMDS.CONVERT,
      }),
      callback,
    );
  }

  public convertLimits(
    options: CoinpaymentsConvertLimitsOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsConvertLimitsResponse>,
  ) {
    return request<CoinpaymentsConvertLimitsResponse>(
      this.credentials,
      mapPayload<CoinpaymentsConvertLimitsOpts>(options, {
        cmd: CMDS.CONVERT_LIMITS,
      }),
      callback,
    );
  }

  public getWithdrawalHistory(
    options?:
      | CoinpaymentsGetWithdrawalHistoryOpts
      | CoinpaymentsReturnCallback<CoinpaymentsGetWithdrawalHistoryResponse>,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsGetWithdrawalHistoryResponse>,
  ) {
    if (typeof options === 'function') {
      callback = options;
      options = {};
    }
    return request<CoinpaymentsGetWithdrawalHistoryResponse>(
      this.credentials,
      mapPayload<CoinpaymentsGetWithdrawalHistoryOpts>(options, {
        cmd: CMDS.GET_WITHDRAWAL_HISTORY,
      }),
      callback,
    );
  }

  public getConversionInfo(
    options: CoinpaymentsGetConversionInfoOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsConversionInfoResponse>,
  ) {
    return request<CoinpaymentsConversionInfoResponse>(
      this.credentials,
      mapPayload<CoinpaymentsGetConversionInfoOpts>(options, {
        cmd: CMDS.GET_CONVERSATION_INFO,
      }),
      callback,
    );
  }

  public getProfile(
    options: CoinpaymentsGetProfileOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsGetProfileResponse>,
  ) {
    return request<CoinpaymentsGetProfileResponse>(
      this.credentials,
      mapPayload<CoinpaymentsGetProfileOpts>(options, {
        cmd: CMDS.GET_TAG_INFO,
      }),
      callback,
    );
  }

  public tagList(
    callback?: CoinpaymentsReturnCallback<CoinpaymentsTagListResponse>,
  ) {
    return request<CoinpaymentsTagListResponse>(
      this.credentials,
      mapPayload<CoinpaymentsTagListOpts>(
        {},
        {
          cmd: CMDS.GET_TAG_LIST,
        },
      ),
      callback,
    );
  }

  public updateTagProfile(
    options: CoinpaymentsUpdateTagProfileOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsUpdateTagProfileResponse>,
  ) {
    return request<CoinpaymentsUpdateTagProfileResponse>(
      this.credentials,
      mapPayload<CoinpaymentsUpdateTagProfileOpts>(options, {
        cmd: CMDS.UPDATE_TAG,
      }),
      callback,
    );
  }

  public claimTag(
    options: CoinpaymentsClaimTagOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsClaimTagResponse>,
  ) {
    return request<CoinpaymentsClaimTagResponse>(
      this.credentials,
      mapPayload<CoinpaymentsClaimTagOpts>(options, {
        cmd: CMDS.CLAIM_TAG,
      }),
      callback,
    );
  }

  public renewTag(
    options: CoinpaymentsRenewTagOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsRenameTagResponse>,
  ) {
    return request<CoinpaymentsRenameTagResponse>(
      this.credentials,
      mapPayload<CoinpaymentsRenewTagOpts>(options, {
        cmd: CMDS.RENEW_TAG,
      }),
      callback,
    );
  }

  public deleteTag(
    options: CoinpaymentsDeleteTagOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsDeleteTagResponse>,
  ) {
    return request<CoinpaymentsDeleteTagResponse>(
      this.credentials,
      mapPayload<CoinpaymentsDeleteTagOpts>(options, {
        cmd: CMDS.DELETE_TAG,
      }),
      callback,
    );
  }

  public claimCoupon(
    options: CoinpaymentsClaimCouponOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsClaimCouponResponse>,
  ) {
    return request<CoinpaymentsClaimCouponResponse>(
      this.credentials,
      mapPayload<CoinpaymentsClaimCouponOpts>(options, {
        cmd: CMDS.CLAIM_COUPON,
      }),
      callback,
    );
  }

  public buyTag(
    options: CoinpaymentsBuyTagOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsBuyTagResponse>,
  ) {
    return request<CoinpaymentsBuyTagResponse>(
      this.credentials,
      mapPayload<CoinpaymentsBuyTagOpts>(options, {
        cmd: CMDS.BUY_TAG,
      }),
      callback,
    );
  }
}

module.exports = Coinpayments;
module.exports.default = Coinpayments;
module.exports.Coinpayments = Coinpayments;

export { Coinpayments };
export default Coinpayments;
