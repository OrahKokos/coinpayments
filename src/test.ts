import { Coinpayments } from './main'

const creds = {
  key: 'key',
  secret: 'secret',
}

const cp = new Coinpayments(creds)
