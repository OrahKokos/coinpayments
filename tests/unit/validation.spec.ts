import {
  getValidationSchema,
  getVariations,
  validateMassWithDrawal,
  hasOneOfFields,
  validateMassWithdrawalOpts,
  filterMassWithdrawalOpts,
} from '../../src/validation';

import { CMDS } from '../../src/constants';

describe('Validation unit tests', () => {
  describe('getVariations tests', () => {
    it('Should return array on no variation', () => {
      const NO_VARIATION = ['prop1', 'prop2', 'prop3'];
      const variations = getVariations(NO_VARIATION);
      expect(variations).toBeInstanceOf(Array);
      expect(variations.length).toBe(1);
      expect(variations[0]).toStrictEqual(NO_VARIATION);
      expect(variations).toMatchSnapshot();
    });
    it('Should return array on variation', () => {
      const VARIATION = ['prop1', 'prop2', 'prop3', ['prop4a', 'prop4b']];
      const EXPECTED_VARIATION_1 = ['prop1', 'prop2', 'prop3', 'prop4a'];
      const EXPECTED_VARIATION_2 = ['prop1', 'prop2', 'prop3', 'prop4b'];
      const variations = getVariations(VARIATION);
      expect(variations).toBeInstanceOf(Array);
      expect(variations.length).toBe(2);
      expect(variations[0]).toBeInstanceOf(Array);
      expect(variations[0]).toStrictEqual(EXPECTED_VARIATION_1);
      expect(variations[1]).toBeInstanceOf(Array);
      expect(variations[1]).toStrictEqual(EXPECTED_VARIATION_2);
      expect(variations).toMatchSnapshot();
    });
  });
  describe('validateMassWithDrawal tests', () => {
    it('Should return false on invalid payload', () => {
      const INVALID_PAYLOAD = {
        cmd: CMDS.CREATE_MASS_WITHDRAWAL,
        'wd[wd1][amount]': 1.0,
        'wd[wd1][address]': '1BitcoinAddress',
        'wd[wd1][currency]': 'BTC',
        'wd[wd2][amount]': 0.0001,
        'wd[wd2][address]': '1BitcoinAddress',
        'wd[wdError][currency]': 'BTC',
        'wd[wd3][amount]': 0.0001,
        'wd[wd3][address]': '1BitcoinAddress',
        'wd[wd3][currency]': 'LTC',
        'wd[wd4][amount]': 0.01,
        'wd[wd4][address]': '1BitcoinAddress',
        'wd[wd4][currency]': 'BTC',
      };
      expect(validateMassWithDrawal(INVALID_PAYLOAD)).toBeFalsy();
    });
    it('Should return true on valid payload', () => {
      const VALID_PAYLOAD = {
        cmd: CMDS.CREATE_MASS_WITHDRAWAL,
        'wd[wd1][amount]': 1.0,
        'wd[wd1][address]': '1BitcoinAddress',
        'wd[wd1][currency]': 'BTC',
        'wd[wd2][amount]': 0.0001,
        'wd[wd2][address]': '1BitcoinAddress',
        'wd[wd2][currency]': 'BTC',
        'wd[wd3][amount]': 0.0001,
        'wd[wd3][address]': '1BitcoinAddress',
        'wd[wd3][currency]': 'LTC',
        'wd[wd4][amount]': 0.01,
        'wd[wd4][address]': '1BitcoinAddress',
        'wd[wd4][currency]': 'BTC',
      };
      expect(validateMassWithDrawal(VALID_PAYLOAD)).toBeTruthy();
    });
  });
  describe('getValidationSchema tests', () => {
    it('Should throw error on undefined', () => {
      let done = 0;
      try {
        getValidationSchema(undefined);
      } catch (e) {
        done++;
        expect(e.message).toBe('No method passed');
      }
      expect(done).toBeTruthy();
    });
    it('Should throw error on unknown command', () => {
      const UNKNOWN = 'unknown';
      let done = 0;
      try {
        getValidationSchema(UNKNOWN);
      } catch (e) {
        done++;
        expect(e.message).toBe(`No such method ${UNKNOWN}`);
      }
      expect(done).toBeTruthy();
    });
    it('Should return validation schema', () => {
      Object.keys(CMDS).forEach(cmd => {
        const schema = getValidationSchema(CMDS[cmd]);
        expect(schema).toMatchSnapshot();
      });
    });
  });
  describe('hasOneOfFields', () => {
    it('Should return true on mutually exclusive', () => {
      const object = {
        x: 'x',
        y: 'y',
        z: 'z',
      };
      const keys1 = ['x', 's', 'q'];
      const keys2 = ['s', 'q', 'x'];
      expect(hasOneOfFields(object, keys1)).toBe(true);
      expect(hasOneOfFields(object, keys2)).toBe(true);
    });
    it('Should return false on mutually non-exclusive', () => {
      const object = {
        x: 'x',
        y: 'y',
        z: 'z',
      };
      const keys1 = ['x', 'y', 'q'];
      const keys2 = ['s', 'y', 'z'];
      const keys3 = ['x', 'y', 'z'];
      expect(hasOneOfFields(object, keys1)).toBe(false);
      expect(hasOneOfFields(object, keys2)).toBe(false);
      expect(hasOneOfFields(object, keys3)).toBe(false);
    });
  });
  describe('validateMassWithdrawalOpts', () => {
    it('Should be true on valid payload', () => {
      const validPayload1 = {
        currency: 'BTC',
        amount: 1,
        address: 'x',
      };
      const validPayload2 = {
        currency: 'BTC',
        amount: 1,
        domain: 'x',
      };
      const validPayload3 = {
        currency: 'BTC',
        amount: 1,
        pbntag: 'x',
      };
      expect(validateMassWithdrawalOpts(validPayload1)).toBe(true);
      expect(validateMassWithdrawalOpts(validPayload2)).toBe(true);
      expect(validateMassWithdrawalOpts(validPayload3)).toBe(true);
    });
    it('Should be false on invalid payload', () => {
      const invalidPayload1 = {
        currency: 'BTC',
        amount: 1,
        address: 'x',
        domain: 'x',
      };
      const invalidPayload2 = {
        currency: 'BTC',
        domain: 'x',
        pbntag: 'x',
        amount: 1,
      };
      const invalidPayload3 = {
        currency: 'BTC',
        amount: 1,
        pbntag: 'x',
        address: 'x',
        domain: 'x',
      };
      expect(validateMassWithdrawalOpts(invalidPayload1)).toBe(false);
      expect(validateMassWithdrawalOpts(invalidPayload2)).toBe(false);
      expect(validateMassWithdrawalOpts(invalidPayload3)).toBe(false);
    });
  });
  describe('filterMassWithdrawalOpts', () => {
    it('Should return same array if all payload is valid', () => {
      const validPayload1 = {
        currency: 'BTC',
        amount: 1,
        address: 'x',
      };
      const validPayload2 = {
        currency: 'BTC',
        amount: 1,
        domain: 'x',
      };
      const validPayload3 = {
        currency: 'BTC',
        amount: 1,
        pbntag: 'x',
      };
      const validPayload = [validPayload1, validPayload2, validPayload3];
      const filteredArray = filterMassWithdrawalOpts(validPayload);
      expect(filteredArray.sort()).toEqual(validPayload.sort());
    });
    it('Should filter out invalid payloads', () => {
      const validPayload1 = {
        currency: 'BTC',
        amount: 1,
        address: 'x',
      };
      const validPayload2 = {
        currency: 'BTC',
        amount: 1,
        domain: 'x',
      };
      const validPayload3 = {
        currency: 'BTC',
        amount: 1,
        pbntag: 'x',
        domain: 'x',
      };
      const invalidPayload = [validPayload1, validPayload2, validPayload3];
      const validPayload = [validPayload1, validPayload2];
      const filteredArray = filterMassWithdrawalOpts(invalidPayload);
      expect(filteredArray.sort()).toEqual(validPayload.sort());
    });
  });
});
