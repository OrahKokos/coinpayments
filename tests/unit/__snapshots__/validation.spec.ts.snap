// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validation unit tests getValidationSchema tests Should return validation schema 1`] = `Array []`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 2`] = `Array []`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 3`] = `
Array [
  "amount",
  "currency1",
  "currency2",
  "buyer_email",
]
`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 4`] = `
Array [
  "currency",
]
`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 5`] = `
Array [
  "txid",
]
`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 6`] = `
Array [
  "txid",
]
`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 7`] = `Array []`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 8`] = `Array []`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 9`] = `
Array [
  "currency",
]
`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 10`] = `
Array [
  "amount",
  "currency",
  Array [
    "merchant",
    "pbntag",
  ],
]
`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 11`] = `
Array [
  "amount",
  "currency",
  Array [
    "address",
    "pbntag",
    "domain",
  ],
]
`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 12`] = `Array []`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 13`] = `
Array [
  "id",
]
`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 14`] = `
Array [
  "amount",
  "from",
  "to",
]
`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 15`] = `
Array [
  "from",
  "to",
]
`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 16`] = `Array []`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 17`] = `
Array [
  "id",
]
`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 18`] = `
Array [
  "id",
]
`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 19`] = `
Array [
  "pbntag",
]
`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 20`] = `Array []`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 21`] = `
Array [
  "coin",
]
`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 22`] = `
Array [
  "tagid",
  "name",
]
`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 23`] = `
Array [
  "tagid",
]
`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 24`] = `
Array [
  "tagid",
  "coin",
]
`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 25`] = `
Array [
  "tagid",
]
`;

exports[`Validation unit tests getValidationSchema tests Should return validation schema 26`] = `
Array [
  "coupon",
]
`;

exports[`Validation unit tests getVariations tests Should return array on no variation 1`] = `
Array [
  Array [
    "prop1",
    "prop2",
    "prop3",
  ],
]
`;

exports[`Validation unit tests getVariations tests Should return array on variation 1`] = `
Array [
  Array [
    "prop1",
    "prop2",
    "prop3",
    "prop4a",
  ],
  Array [
    "prop1",
    "prop2",
    "prop3",
    "prop4b",
  ],
]
`;
